x/staking/keeper/unbonding.go

```diff
@@ -329,24 +329,16 @@ func (k Keeper) validatorUnbondingCanComplete(ctx sdk.Context, id uint64) (found
 		return false, nil
 	}
 
-	if !val.IsMature(ctx.BlockTime(), ctx.BlockHeight()) {
-		val.UnbondingOnHold = false
-		k.SetValidator(ctx, val)
-	} else {
```

Comment:
Why did this get removed? Doesn't it remove functionality from this function? Where does 

```
		// If unbonding is mature complete it
		val = k.UnbondingToUnbonded(ctx, val)
		if val.GetDelegatorShares().IsZero() {
			k.RemoveValidator(ctx, val.GetOperator())
		}

		k.DeleteUnbondingIndex(ctx, id)
```

happen now?

Reply:
I see it is also happening in `UnbondAllMatureValidators`, but I'm not sure that's enough, because there it is only triggered when unbonding on the provider completes, while it is trigger here when unbonding on the consumer completes.

Reply:
Seems like the logic might now be broken if unbonding completes on the provider chain first, but I think the unit tests test for that so I'm confused.

Reply:
Huh, looks like they don't. I don't remember why not.

Reply:
@jtremback `UnbondAllMatureValidators` is called in every `staking.EndBlock()`. As long as a mature validator `val` (that is on hold  due to Interchain Security, i.e., `val.UnbondingOnHoldRefCount > 0`) is not removed from the unbonding validator queue, then once `UnbondingCanComplete` is called enough times to decrement `val.UnbondingOnHoldRefCount` to `0`, the next call to `staking.EndBlock()` will move `val` to `unbonded`.

Reply:
Notice that for UBDEs and REDEs, we cannot remove the code from `unbondingDelegationEntryCanComplete()` and `redelegationEntryCanComplete`, respectively. This is because in `staking.EndBlock()` mature unbonding delegations and redelegations are removed from their respective queues, see `DequeueAllMatureUBDQueue` and `DequeueAllMatureRedelegationQueue` [here](https://github.com/cosmos/cosmos-sdk/blob/af6ee93e2923eff159d07384f9b80010e7358b71/x/staking/keeper/val_state_change.go#L18).